File: src/actions_py/actions_py/count_until_server.py
#!/usr/bin/env python3
import time
import enum
import threading
import rclpy
from rclpy.node import Node
from rclpy.action import (
    ActionServer,
    GoalResponse,
    CancelResponse
)
from rclpy.action.server import ServerGoalHandle
from rclpy.executors import MultiThreadedExecutor
from rclpy.callback_groups import ReentrantCallbackGroup
from demo_interfaces.action import CountUntil

# Policy on how to handle multiple action requests
class GoalPolicy(enum.Enum):
    Parralel = 1
    SingleGoal = 2
    PreEmptCurrent = 3
    Queue = 4

class CountUntilServerNode(Node):
    def __init__(self, goal_policy: GoalPolicy):
        super().__init__("count_until_server_node")

        self.goal_policy_ = goal_policy
        self.goal_handle_: ServerGoalHandle = None
        self.goal_lock_ = threading.Lock()

        if self.goal_policy_ == GoalPolicy.Queue:
            self.goal_queue_ = []

        self.count_until_server_ = ActionServer(
            self, 
            CountUntil, 
            "count_until",
            goal_callback=self.goal_callback,
            handle_accepted_callback=self.handle_accepted_callback,
            cancel_callback=self.cancel_callback,
            execute_callback=self.execute_callback,
            callback_group=ReentrantCallbackGroup())
        
    def goal_callback(self, goal_request: CountUntil.Goal):
        self.get_logger().info("Received a goal")

        # Policy: reject new goal if current goal is active
        if self.goal_policy_ == GoalPolicy.SingleGoal:
            with self.goal_lock_:   
                if self.goal_handle_ is not None and \
                    self.goal_handle_.is_active:

                    self.get_logger().info("A goal is already acrive, rejecting new goal")
                    return GoalResponse.REJECT

        # Validate the goal request
        if goal_request.target_number <= 0:
            self.get_logger().info("Reject the goal: number is: " + \
                str(goal_request.target_number))
            return GoalResponse.REJECT
        
        # Policy pre-empt current goal if new goal requested
        if self.goal_policy_ == GoalPolicy.PreEmptCurrent:
            with self.goal_lock_:
                if self.goal_handle_ is not None and \
                    self.goal_handle_.is_active:

                    self.get_logger().info("Abort current goal, accept new goal")
                    self.goal_handle_.abort()


        self.get_logger().info("Accepting the goal: number is: " + \
            str(goal_request.target_number))
        return GoalResponse.ACCEPT

    def handle_accepted_callback(self, goal_handle: ServerGoalHandle):
        if self.goal_policy_ != GoalPolicy.Queue:
            goal_handle.execute()
            return

        with self.goal_lock_:
            if self.goal_handle_ is not None:
                self.goal_queue_.append(goal_handle)
            else:
                goal_handle.execute()

    def cancel_callback(self, goal_handle: ServerGoalHandle):
        self.get_logger().info("Received a cencel request")
        return CancelResponse.ACCEPT

    def execute_callback(self, goal_handle: ServerGoalHandle):
        with self.goal_lock_:
            self.goal_handle_ = goal_handle

        # Get request from the goal
        target_number = goal_handle.request.target_number
        period = goal_handle.request.period

        # Execute the action
        self.get_logger().info("Executing the goal")
        feedback = CountUntil.Feedback()
        
        counter = 0
        for _ in range (target_number):
            if not goal_handle.is_active:
                return self.trigger_action_end(counter=counter)
            
            if goal_handle.is_cancel_requested:
                self.get_logger().info("Cancel: stopping action execution")
                goal_handle.canceled()
                return self.trigger_action_end(counter=counter)
            
            if should_abort(counter=counter):
                goal_handle.abort()
                return self.trigger_action_end(counter=counter)
            
            counter += 1
            feedback.current_number = counter
            goal_handle.publish_feedback(feedback=feedback)
            time.sleep(period)

        # Set goal final state
        if counter == target_number:
            goal_handle.succeed()

        # Send the result
        return self.trigger_action_end(counter=counter)
    
    def trigger_action_end(self, counter) -> CountUntil.Result:
        result = CountUntil.Result()
        result.reached_number = counter
        self.process_next_goal_in_queue()
        return result

    def process_next_goal_in_queue(self):
        if not self.goal_policy_ == GoalPolicy.Queue:
            return

        with self.goal_lock_:
            if self.goal_queue_:
                self.goal_queue_.pop(0).execute()
            else:
                self.goal_handle_ = None

def should_abort(counter):
    # return counter > 0 and counter %3 == 0
    return False

def main(args=None):
    rclpy.init(args=args)
    node = CountUntilServerNode(GoalPolicy.Queue)
    rclpy.spin(node, MultiThreadedExecutor())
    rclpy.shutdown()


if __name__ == "__main__":
    main()

File: src/actions_cpp/src/count_until_server.cpp
#include <queue>
#include <memory>
#include <thread>

#include "actions_cpp/count_until_action_includes.hpp"
#include "actions_cpp/goal_policy.hpp"
using CountUntilGoalHandle = rclcpp_action::ServerGoalHandle<CountUntil>;

class NumberLoop {
public:
    NumberLoop(int target) : 
        target_(target) {
            std::cout<< target_;
        }

    void tick() {
        if (counter_ < target_) {
            counter_++;
            std::cout << counter_;
        }
    }

    bool is_finished() const {
        return counter_ == target_;
    }

    int get_counter() const {
        return counter_;
    }

private:
    int counter_ {0};
    int target_ {0};
};

bool goal_valid(int target_number) {
    return target_number >= 0;
}

class CountUntilServerNode : public rclcpp::Node
{
public:
    CountUntilServerNode(GoalPolicy policy) : Node("count_until_server_node")
    {
        policy_ = policy;
        if (policy_ == GoalPolicy::Queue) {
            goals_worker_ = std::thread(create_execute_callback_queue());
        }

        callback_group_ = create_callback_group(rclcpp::CallbackGroupType::Reentrant);

        server_ = rclcpp_action::create_server<CountUntil>(
            this,
            count_until_action_name,
            create_goal_callback(),
            create_cancel_callback(),
            create_handle_accepted_callback(),
            rcl_action_server_get_default_options(),
            callback_group_
        );
        RCLCPP_INFO(get_logger(), "Action server has started");
    }

    ~CountUntilServerNode() {
        goals_worker_.join();
    }

private:
    rclcpp_action::Server<CountUntil>::SharedPtr server_;
    rclcpp::CallbackGroup::SharedPtr callback_group_;
    GoalPolicy policy_;
    
    std::queue<std::shared_ptr<CountUntilGoalHandle>> goals_;
    std::mutex goals_lock_;
    std::thread goals_worker_;

    std::function<void()> create_execute_callback_queue() 
    {
        return [this]() {
            rclcpp::Rate loop_rate(10);
            while(rclcpp::ok()) {
                std::shared_ptr<CountUntilGoalHandle> next_goal;
                {
                    std::lock_guard<std::mutex> guard(goals_lock_);
                    if (!goals_.empty())
                    {
                        next_goal = goals_.front();
                        goals_.pop();
                    }                    
                }

                if (next_goal) {
                    RCLCPP_INFO(get_logger(), "Executing next goal in queue");
                    execute_goal(next_goal);
                }

                loop_rate.sleep();
            }
        };
    } 

    std::function<rclcpp_action::GoalResponse(
        const rclcpp_action::GoalUUID&,
        std::shared_ptr<const CountUntil::Goal>)> create_goal_callback() 
    {
        return [this](const rclcpp_action::GoalUUID &/*uuid*/,
                            std::shared_ptr<const CountUntil::Goal> goal)
        {
            RCLCPP_INFO(get_logger(), "Received a goal");

            switch (policy_) 
            {
                case GoalPolicy::Parallel:
                    break;
                case GoalPolicy::RejectSecond:
                {
                    if (goal_running()) {
                        RCLCPP_INFO(get_logger(), "Rejected a goal, current goal still running");
                        return rclcpp_action::GoalResponse::REJECT;
                    }
                    break;
                }
                case GoalPolicy::OverrideFirst:
                {
                    if(goal_running() && 
                        goal_valid(goal->target_number)) 
                    {
                        std::lock_guard<std::mutex> guard(goals_lock_);
                        // Note: could also abort in the execute loop by remembering the UUID
                        goals_.front()->abort(generate_result(-1));
                        goals_.pop();
                        RCLCPP_INFO(get_logger(), "Overriding current goal");
                    }
                    break;
                }
                case GoalPolicy::Queue:
                    RCLCPP_INFO(get_logger(), "Adding goal to queue");
                    break;
            }

            if (!goal_valid(goal->target_number)) 
            {
                RCLCPP_INFO(get_logger(), "Rejected a goal");
                return rclcpp_action::GoalResponse::REJECT;
            }
            
            RCLCPP_INFO(get_logger(), "Accepted a goal");
            return rclcpp_action::GoalResponse::ACCEPT_AND_EXECUTE;
        };
    }

    std::function<rclcpp_action::CancelResponse(
        const std::shared_ptr<CountUntilGoalHandle>)> create_cancel_callback()
    {
        return [this](
            const std::shared_ptr<CountUntilGoalHandle> /*goal_handle*/)
        {
            RCLCPP_INFO(get_logger(), "Received a cancel request");
            RCLCPP_INFO(get_logger(), "Canceling the goal");
            return rclcpp_action::CancelResponse::ACCEPT;
        };
    }

    std::function<void(
        const std::shared_ptr<CountUntilGoalHandle>)> 
        create_handle_accepted_callback()
    {
        return [this](
            const std::shared_ptr<CountUntilGoalHandle> goal_handle)
        {
            add_goal_to_queue(goal_handle);
            if (policy_ != GoalPolicy::Queue) {
                execute_goal(goal_handle);
            }
        };
    }

    void execute_goal(const std::shared_ptr<CountUntilGoalHandle> goal_handle)
    {
        NumberLoop number_loop(goal_handle->get_goal()->target_number);

        float sleep_time_seconds = goal_handle->get_goal()->period;
        rclcpp::Rate loop_rate(1.0 / sleep_time_seconds);

        auto feedback = std::make_shared<CountUntil::Feedback>();
        while(! number_loop.is_finished()) {
            if (goal_handle->is_canceling()) {
                goal_handle->canceled(
                    generate_result(number_loop.get_counter()));
                    RCLCPP_INFO(get_logger(), "Goal was canceled");
                return;
            }

            if (!goal_handle->is_active()) {
                generate_result(number_loop.get_counter());
                    RCLCPP_INFO(get_logger(), "Goal not active");
                return;
            }

            number_loop.tick();
            feedback->current_number = number_loop.get_counter();
            goal_handle->publish_feedback(feedback);

            loop_rate.sleep();
        }

        goal_handle->succeed(
            generate_result(number_loop.get_counter()));
    }

    std::shared_ptr<CountUntilResult> generate_result(int counter)
    {
        auto result = std::make_shared<CountUntil::Result>();
        result->reached_number = counter;
        return result;
    }

    bool goal_running() {
        std::lock_guard<std::mutex> guard(goals_lock_);
        return ! goals_.empty() &&
        goals_.front()->is_active();
    }

    void add_goal_to_queue(const std::shared_ptr<CountUntilGoalHandle> goal) {
        std::lock_guard<std::mutex> guard(goals_lock_);
        goals_.push(goal);
    }
};

int main(int argc, char **argv)
{
    rclcpp::init(argc, argv);
    auto node = std::make_shared<CountUntilServerNode>(GoalPolicy::Queue);

    rclcpp::executors::MultiThreadedExecutor executor;
    executor.add_node(node);
    executor.spin();
    rclcpp::shutdown();
    return 0;
}
